packages/jsii-pacmak/lib/targets/python/type-name.ts

```diff
@@ -0,0 +1,354 @@
+
+  return optional ? new Optional(result) : result;
+}
+
+export function mergePythonImports(...pythonImports: readonly PythonImports[]): PythonImports {
+  const result: Record<string, Set<string>> = {};
+  for (const bag of pythonImports) {
+    for (const [packageName, items] of Object.entries(bag)) {
+      if (!(packageName in result)) {
+        result[packageName] = new Set();
+      }
+      for (const item of items) {
+        result[packageName].add(item);
+      }
+    }
+  }
+  return result;
+}
+
+function isOptionalValue(type: OptionalValue | TypeReference): type is OptionalValue {
+  return (type as unknown as OptionalValue).type != null;
+}
+
+class Dict implements TypeName {
+  // eslint-disable-next-line @typescript-eslint/explicit-member-accessibility
+  readonly #element: TypeName;
+
+  public constructor(element: TypeName) {
+    this.#element = element;
+  }
+
+  public pythonType(context: NamingContext) {
+    return `typing.Mapping[str,${this.#element.pythonType(context)}]`;
+  }
+
+  public requiredImports(context: NamingContext) {
+    return mergePythonImports(this.#element.requiredImports(context), { typing: new Set(['']) });
+  }
+}
+
+class List implements TypeName {
+  // eslint-disable-next-line @typescript-eslint/explicit-member-accessibility
+  readonly #element: TypeName;
+
+  public constructor(element: TypeName) {
+    this.#element = element;
+  }
+
+  public pythonType(context: NamingContext) {
+    return `typing.List[${this.#element.pythonType(context)}]`;
+  }
+
+  public requiredImports(context: NamingContext) {
+    return mergePythonImports(this.#element.requiredImports(context), { typing: new Set(['']) });
+  }
+}
+
+class Optional implements TypeName {
+  // eslint-disable-next-line @typescript-eslint/explicit-member-accessibility
+  readonly #wrapped: TypeName;
+
+  public constructor(wrapped: TypeName) {
+    this.#wrapped = wrapped;
+  }
+
+  public pythonType(context: NamingContext) {
+    const optionalType = this.#wrapped.pythonType({ ...context, ignoreOptional: true });
+    if (context.ignoreOptional || this.#wrapped === Primitive.ANY) {
+      return optionalType;
+    }
+    return `typing.Optional[${optionalType}]`;
+  }
+
+  public requiredImports(context: NamingContext) {
+    const wrapped = this.#wrapped.requiredImports({ ...context, ignoreOptional: true });
+    if (context.ignoreOptional) {
+      return wrapped;
+    }
+    return mergePythonImports(wrapped, { typing: new Set(['']) });
+  }
+}
+
+class Primitive implements TypeName {
+  private static readonly BOOL = new Primitive('bool', {});
+  private static readonly DATE = new Primitive('datetime.datetime', { 'datetime': new Set(['']) });
+  private static readonly JSII_NUMBER = new Primitive('jsii.Number', {}); // "jsii" is always already imported!
+  private static readonly STR = new Primitive('str', {});
+  private static readonly JSON = new Primitive('typing.Mapping[typing.Any, typing.Any]', { 'typing': new Set(['']) });
```

Comment:
These private statics feel unnecessary... I would just inline them inside `of()`.

Reply:
They help reduce churn. Inlining in `of` would take more work that would hurt the overall performance (and I'd argue wouldn't improve readability).